Add type functions (#307)
axiomhq/docs: apl/scalar-functions/type-functions/isreal.mdx

---
title: isreal
description: 'This page explains how to use the isreal function in APL.'
---

> Use the `isreal` function to determine whether a value is a real (non-NaN) numeric value. This function is helpful when you need to validate data before performing numeric operations. For example, you can use `isreal` to filter out invalid values such as NaN (Not a Number), which could otherwise disrupt aggregations or calculations.

Reviewer: `real` just means that it's a float type in computing real/float are essentially interchangeable. as such `NaN/Inf` are both real numbers too because they are also float values. so this line needs to change. 

we have `isnan()` and `isinf()` to check for nan/inf values that are stored in real values which is worth highlighting

Author: @gordallott Thanks a lot! Updated

> Use the `isreal` function to determine whether a value is a real (non-NaN) numeric value. This function is helpful when you need to validate data before performing numeric operations. For example, you can use `isreal` to filter out invalid values such as NaN (Not a Number), which could otherwise disrupt aggregations or calculations.

Author: ```suggestion
Use the `isreal` function to determine whether a value is a real number. This function is helpful when you need to validate data before performing numeric operations. For example, you can use `isreal` to filter out invalid values which could otherwise disrupt aggregations or calculations.
```

You often use `isreal` in data cleaning pipelines, conditional logic, and when inspecting metrics like durations, latencies, or numeric identifiers. It’s especially useful when working with telemetry or log data that includes optional or incomplete numeric fields.

## For users of other query languages

If you come from other query languages, this section explains how to adjust your existing queries to achieve the same results in APL.

<AccordionGroup>
<Accordion title="Splunk SPL users">

> Splunk uses the `isnum` function to check whether a string represents a numeric value. However, `isnum` does not verify that the value is finite and not NaN. In APL, `isreal` ensures that the value is a valid, real number and not NaN.

Author: ```suggestion
Splunk uses the `isnum` function to check whether a string represents a numeric value.
```

<CodeGroup>
```sql Splunk example
... | eval is_valid = if(isnum(duration), "yes", "no")
````

```kusto APL equivalent
... | extend is_valid = iff(isreal(duration), 'yes', 'no')
```

</CodeGroup>

</Accordion>
<Accordion title="ANSI SQL users">

> ANSI SQL does not have a direct equivalent to `isreal`. You typically check for numeric values using `IS NOT NULL` and avoid known invalid markers manually. APL’s `isreal` abstracts this by directly checking if a value is a finite, non-NaN number.

Author: ```suggestion
ANSI SQL does not have a direct equivalent to `isreal`. You typically check for numeric values using `IS NOT NULL` and avoid known invalid markers manually. APL’s `isreal` abstracts this by directly checking if a value is a real number.
```

<CodeGroup>
```sql SQL example
SELECT *, 
       CASE WHEN duration IS NOT NULL THEN 'yes' ELSE 'no' END AS is_valid
FROM traces
```

```kusto APL equivalent
['otel-demo-traces']
| extend is_valid = iff(isreal(duration), 'yes', 'no')
```

</CodeGroup>

</Accordion>
</AccordionGroup>

## Usage

### Syntax

```kusto
isreal(value)
```

### Parameters

| Name  | Type    | Description                  |
| ----- | ------- | ---------------------------- |
| value | any | The input value to evaluate. |

### Returns

> Returns `true` if the input is a valid real number and not NaN. Returns `false` for NaN, strings, nulls, or non-numeric types.

Author: ```suggestion
Returns `true` if the input is a valid real number. Returns `false` for strings, nulls, or non-numeric types.
```
